pulsar-broker/src/main/java/org/apache/pulsar/broker/service/RedeliveryTracker.java

```diff
@@ -19,17 +19,25 @@
 package org.apache.pulsar.broker.service;
 
 import java.util.List;
+import java.util.function.Supplier;
+import org.apache.bookkeeper.mledger.Entry;
 import org.apache.bookkeeper.mledger.Position;
 
 public interface RedeliveryTracker {
 
-    int incrementAndGetRedeliveryCount(Position position);
+    int incrementAndGetRedeliveryCount(Position position, Consumer consumer);
 
     int getRedeliveryCount(Position position);
 
-    void remove(Position position);
+    void remove(Position position, Position markDeletedPosition);
 
-    void removeBatch(List<Position> positions);
+    void removeBatch(List<Position> positions, Position markDeletedPosition);
 
     void clear();
+
+    default void noticeConsumerClosed(Consumer consumer){}
+
+    default Consumer cherryNextConsumer(List<Entry> entries, Supplier<Consumer> nextConsumerFunc, int consumerCount){
```

Comment:
what about "pickNextConsumer", "cherry" doesn't sound good to me

Reply:
ah, good.  Already fixed